lib/puppet/confine/boolean.rb

```diff
@@ -0,0 +1,56 @@
+require 'puppet/confine'
+
+# Common module for the Boolean confines. It currently
+# contains just enough code to implement PUP-9336.
+class Puppet::Confine
+  module Boolean
+    # Returns the passing value for the Boolean confine. This should
+    # just be the class name.
+    def passing_value
+      return @passing_value unless @passing_value.nil?
+
+      class_name = self.class.to_s.split('::').pop.downcase.to_sym
+
+      @passing_value = class_name == :true ? true : false
+    end
+
+    # The Boolean confines 'true' and 'false' let the user specify
+    # two types of values:
+    #     * A lambda for lazy evaluation. This would be something like
+    #         confine :true => lambda { true }
+    #
+    #     * A single Boolean value, or an array of Boolean values. This would
+    #     be something like
+    #         confine :true => true OR confine :true => [true, false, false, true]
+    #
+    # This override distinguishes between the two cases.
+    def values
+      # Note that Puppet::Confine's constructor ensures that @values
+      # will always be an array, even if a lambda's passed in. This is
+      # why we have the length == 1 check.
+      unless @values.length == 1 && @values.first.respond_to?(:call)
+        return @values
+      end
+
+      # We have a lambda. Here, we want to enforce "cache positive"
+      # behavior, which is to cache the result _if_ it evaluates to
+      # the passing value (i.e. the class name).
+
+      return @cached_value unless @cached_value.nil?
+
+      # Double negate to coerce the value into a Boolean
+      calculated_value = !! @values.first.call
+      if calculated_value == passing_value
+        @cached_value = [calculated_value]
+      end
+
+      [calculated_value]
+    end
+
+    def reset
+      # Reset the cache.  We want to cache it during a given
```

Comment:
Is this necessary? Got this off of https://github.com/ekinanp/puppet/blob/PUP-9336/lib/puppet/confine/variable.rb#L38-L42

Reply:
I think it is, but it also looks like confinement caching is basically worthless (though I'd be happy to be proven wrong here).

`reset` is called whenever we check if a confine matches: https://github.com/puppetlabs/puppet/blob/16220673acaa03f708f6b6410a659b4ba737ab7a/lib/puppet/confine.rb#L66-L78

We check if a confine maches whenever we check if a resource is suitable: https://github.com/puppetlabs/puppet/blob/16220673acaa03f708f6b6410a659b4ba737ab7a/lib/puppet/confiner.rb#L43-L45

We check if a resource is suitable as we come across each resource while we're traversing the relationship graph: https://github.com/puppetlabs/puppet/blob/16220673acaa03f708f6b6410a659b4ba737ab7a/lib/puppet/graph/relationship_graph.rb#L115-L116

Soooo... it looks like attempting to cache anything in the confines is a futile effort as things are now. ☹️  Sorry I didn't notice this sooner.

Reply:
Oh, I didn't know about the "put a link on a new line and it will show you the code" feature. That's cool.

I think the caching is still useful isn't it? The `suitable?` method in `confiner.rb` is a class-level method for the providers based on what I'm seeing in:
https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider.rb#L46

Reply:
The main point I was getting at is that there's two things going on (only the first of which is actually desirable):
1) We check suitability for every resource we run across in the relationship graph.
2) The act of checking provider suitability destroys the cache according to the API as it currently is, because the `ensure` in the `valid?` method defined in `confine.rb` will always call `reset`.

Because of the second point there, the act of checking "is this suitable" is going to destroy the cache as the last step of answering the question, instead of leaving it around to be available to respond to future interrogations.

Reply:
For documentation (and me to remember later on), removing the `reset` method in this PR would get us close to the desired caching behavior for lambdas, but the underlying issue still needs to be fixed and is captured in https://tickets.puppetlabs.com/browse/PUP-9371.